File: Source/DotNET/Backend/Dolittle/EventHorizons.cs
// Copyright (c) Dolittle. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

using System.Collections.Generic;
using System.IO;
using Dolittle.SDK.Tenancy;
using Newtonsoft.Json;

namespace Dolittle.Vanir.Backend.Dolittle
{
    public class EventHorizons : Dictionary<TenantId, EventHorizonDefinition[]>
    {
        public static EventHorizons Load()
        {
            var file = Path.Combine(Directory.GetCurrentDirectory(), ".dolittle", "event-horizons.json");
            if (File.Exists(file))
            {
                var json = File.ReadAllText(file);
                return JsonConvert.DeserializeObject<EventHorizons>(json);
            }
            return new EventHorizons();
        }
    }
}

File: Source/typescript/backend/dolittle/EventHorizons.ts
// Copyright (c) Dolittle. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

import fs from 'fs';
import { EventHorizon } from './EventHorizon';
import path from 'path';
import { TenantId } from '@dolittle/sdk.execution';

type EventHorizonsDefinition = { [key: string]: EventHorizon[] }

export class EventHorizons extends Map<TenantId, EventHorizon[]> {
    private static readonly _pathToFile = path.join(process.cwd(), '.dolittle', 'event-horizons.json');

    constructor(eventHorizons: EventHorizonsDefinition) {
        super();

        for (const tenant of Object.keys(eventHorizons)) {
            this.set(TenantId.from(tenant), eventHorizons[tenant]);
        }
    }

    static load(): EventHorizons {
        if (fs.existsSync(EventHorizons._pathToFile)) {
            const json = fs.readFileSync(EventHorizons._pathToFile).toString();
            const definition = JSON.parse(json) as EventHorizonsDefinition;
            return new EventHorizons(definition);
        }

        return new EventHorizons({});

    }
}

